docs/rfcs/0003-rfc-controller_refactor.md

```diff
@@ -0,0 +1,111 @@
+
+Beyond the connectivity checks, the Controller also manages how much time has elapsed since the VPN was toggled on. Having this logic in the controller bloats the code without any gains, so that code will be moved out of `Controller` and instead a simple `Timer` object will encapsulates the related logic. The timer starts when the VPN enters State On since doing the initial checks may take a few seconds that we don't need to account for.
+
+Because this is all part of a unified connectivity experience, we do not want to release bits and pieces of this until everything is implemented, tested, and ready to go, this means that all implementation will live behind a feature flag until we are ready to test and release.
+
+
+## Implementation Plan
+Introduce a new `ConnectionManager` component that is activated when the client is launched and connects to `networkWatcher` in case anything changes. This new component will manage the following work:
+- Internet probe: assess if there is a network connection.
+- Firewall: assess if firewall is limiting network connectivity.
+- Captive portal check: determine uf the network is behind a captive portal.
+- Server probe: upon activation and then priodically check if there are available servers within user's selection location.
+- Subscription status: ensure that user has an active subscription, and if not, show the "No Subscription Found" screen.
+
+The logic to calculate elapsed time since the VPN was activated will also be moved out of the Controller and into a new `Timer` component. This bit of work will likely require a separate RFC. 
+
+
+The `ConnectionManager` will have two state machines:
+1. `VPNClientState` to determine if the VPN client is off, connecting, or on. 
+2. `ConnectionState` to keep track of which connection related task the VPN is performing at any given point. 
+
+Before the VPN is toggled on, it is in `VPNClientStateOff`. When user toggles on the VPN, the client goes into `VPNClientStateConnecting`, the connection manager will perform the needed connection checks such as `ConnectionStateInternetProbe` and `ConnectionStateServerProbe`. If they all succeed, the VPN will enter `VPNClientStateOn` and call `Controller::activate()`.
+
+The `ConnectionManager` will communicate to the `Controller` to `activate()` and `deactivate()` the VPN. During one of the iterations of this proposal, someone asked if we can have signals instead of states in the `ConnectionManager`. The short answer is no. While emitting a signal to notify the UI about what we should be showing the user does the same thing, we still need to keep track of the current state internally so we know what to do next in case anything goes wrong at any other point.
```

Comment:
Why is it that we need to keep track of the state internally? How does it change what we are going to do next? That is unclear to me on the `ConnectionState` specifically.

Reply:
These are asynchronous tasks that may trigger both upon activation && while the VPN is already active. For instance there is no point in probing servers or checking subscription if network probe is failing but we need a mechanism to keep track of it.

Reply:
Ok, but where do we store the state about whether network probe failed or not? The description of `ConnectionState` says

>  [...] `ConnectionState` to keep track of which connection related task the VPN is performing at any given point via the corresponding task manager `ConnectionTaskManager`. 

So this is just stating which task is being performed at each moment, not the result of the task itself. Is that incorrect?

Reply:
@brizental There would be corresponding bool state variables which would be set to false if a certain probe fails, and we would then emit a signal to the front end stating the failure.

Reply:
I see. In that case, it remains unclear why exactly the ConnectionState is necessary. We can work it out during implementation, you have thought about this longer than I have and it's likely just hard for me to visualize it.